src/main/java/ua/team3/carsharingservice/service/impl/RentalServiceImpl.java

```diff
@@ -0,0 +1,129 @@
+                .stream()
+                .map(rentalMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public RentalDto getById(Long id, User user) {
+        Rental rental = getRentalByIdForUser(id, user);
+        return rentalMapper.toDto(rental);
+    }
+
+    @Override
+    @Transactional
+    public RentalDto create(RentalRequestDto rentalDto, User user) {
+        validateRentalPeriod(rentalDto);
+        validateRentalPermissionFor(user);
+
+        Car car = getCarById(rentalDto.getCarId());
+        decreaseInventoryInCar(car);
+
+        Rental rental = rentalMapper.toModel(rentalDto);
+        rental.setUser(user);
+        rental.setCar(car);
+        Rental savedRental = rentalRepository.save(rental);
+        return rentalMapper.toDto(savedRental);
+    }
+
+    private void decreaseInventoryInCar(Car car) {
+        int availableCarInventory = car.getInventory();
+        if (availableCarInventory < NUM_RENTAL_CARS) {
+            throw new NoCarsAvailableException(
+                    "There are no available cars of this type, please choose another"
+            );
+        }
+        car.setInventory(availableCarInventory - NUM_RENTAL_CARS);
+    }
+
+    @Override
+    @Transactional
+    public RentalDto returnRental(Long id, User user) {
+        Rental rental = getRentalByIdForUser(id, user);
+        ensureRentalNotReturned(rental);
+
+        rental.setActualReturnDate(LocalDate.now());
+        rental.getCar().setInventory(rental.getCar().getInventory() + NUM_RENTAL_CARS);
+        Rental savedRental = rentalRepository.save(rental);
+        return rentalMapper.toDto(savedRental);
+    }
+
+    private void ensureRentalNotReturned(Rental rental) {
+        if (rental.getActualReturnDate() != null) {
+            throw new RentalAlreadyReturnedException(
+                    "The rental with ID: " + rental.getId() + " has already been returned"
+            );
+        }
+    }
+
+    private void validateRentalPeriod(RentalRequestDto dto) {
+        if (dto.getRentalDate().isBefore(LocalDate.now())) {
+            throw new NotValidRentalDateException("Rental date can't be in past");
+        }
+        if (dto.getReturnDate().isBefore(dto.getRentalDate())) {
+            throw new NotValidReturnDateException("Return date should be after rental date");
+        }
+    }
+
+    private void validateRentalPermissionFor(User user) {
+        checkOverdueRentalsFor(user);
+    }
+
+    private void checkOverdueRentalsFor(User user) {
+        boolean hasOverdueRentals = rentalRepository.findByUserId(user.getId()).stream()
+                .anyMatch(rental -> rental.getActualReturnDate() == null
+                        && rental.getReturnDate().isBefore(LocalDate.now()));
+        if (hasOverdueRentals) {
+            throw new NoCarsAvailableException(
+                    "The user has overdue car rentals"
+            );
+        }
+    }
```

Comment:
This will need to be modified later to check if there are any unpaid fine payments